core/src/mast/mod.rs

```diff
@@ -0,0 +1,159 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+use miden_crypto::hash::rpo::RpoDigest;
+
+mod node;
+pub use node::{
+    get_span_op_group_count, BasicBlockNode, CallNode, DynNode, JoinNode, LoopNode, MastNode,
+    OpBatch, SplitNode, OP_BATCH_SIZE, OP_GROUP_SIZE,
+};
+
+use crate::Kernel;
+
+#[cfg(test)]
+mod tests;
+
+pub trait MerkleTreeNode {
+    fn digest(&self) -> RpoDigest;
+    fn to_display<'a>(&'a self, mast_forest: &'a MastForest) -> impl fmt::Display + 'a;
+}
+
+/// An opaque handle to a [`MastNode`] in some [`MastForest`]. It is the responsibility of the user
+/// to use a given [`MastNodeId`] with the corresponding [`MastForest`].
+///
+/// Note that since a [`MastForest`] enforces the invariant that equal [`MastNode`]s MUST have equal
+/// [`MastNodeId`]s, [`MastNodeId`] equality can be used to determine equality of the underlying
+/// [`MastNode`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct MastNodeId(u32);
+
+impl fmt::Display for MastNodeId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MastNodeId({})", self.0)
+    }
+}
+
+// MAST FOREST
+// ===============================================================================================
+
+#[derive(Clone, Debug, Default)]
+pub struct MastForest {
+    /// All of the blocks local to the trees comprising the MAST forest
+    nodes: Vec<MastNode>,
+    node_id_by_hash: BTreeMap<RpoDigest, MastNodeId>,
+
+    /// The "entrypoint", when set, is the root of the entire forest, i.e.
+    /// a path exists from this node to all other roots in the forest. This
+    /// corresponds to the executable entry point. When not set, the forest
+    /// may or may not have such a root in `roots`, but is not required.
+    /// Whether or not the entrypoint is set distinguishes a MAST which is
+    /// executable, versus a MAST which represents a library.
+    ///
+    /// NOTE: The entrypoint is also present in `roots` if set
+    entrypoint: Option<MastNodeId>,
+    kernel: Kernel,
+}
+
+/// Constructors
+impl MastForest {
+    /// Creates a new empty [`MastForest`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+/// Mutators
+impl MastForest {
+    /// Adds a node to the forest, and returns the [`MastNodeId`] associated with it.
+    ///
+    /// If a [`MastNode`] which is equal to the current node was previously added, the previously
+    /// returned [`MastNodeId`] will be returned. This enforces this invariant that equal
+    /// [`MastNode`]s have equal [`MastNodeId`]s.
+    pub fn ensure_node(&mut self, node: MastNode) -> MastNodeId {
+        let node_digest = node.digest();
+
+        if let Some(node_id) = self.node_id_by_hash.get(&node_digest) {
+            // node already exists in the forest; return previously assigned id
+            *node_id
+        } else {
+            let new_node_id =
+                MastNodeId(self.nodes.len().try_into().expect("u32 expected to fit in usize"));
```

Comment:
Changed it to use the `as` operator, so the error message is gone

Reply:
I don't think we want to cast via `as` when going from `usize` to `u32`, only the inverse - as unlikely as it is for us to hit 2^32 nodes in a single `MastForest`, we do want to catch that when inserting new nodes.

Reply:
Looking into it a bit more, `usize` -> `u32` conversions done with the `as` operator are everywhere in the codebase (e.g. [here](https://github.com/0xPolygonMiden/miden-vm/blob/bf885b7bf306d283c23dba44efe0dbd2db11efe6/air/src/constraints/stack/mod.rs#L278)), as well as in winterfell. It's probably best to use `as` here, and use `try_into()` *everywhere* if this is indeed what we want. What do you think?

Reply:
It's safe to use `as` in places where you know that the value in the `usize` is guaranteed to fit in a `u32` (typically because the bounds have already been checked previously, or because the value is derived from something whose bounds are constrained, e.g. the length of a `[T; 64]` array can obviously never overflow `u32`). Since that's not the case here, we definitely want to panic, rather than silently start corrupting programs (the effect here is that we would assign an previously-allocated node id to the new node, but the new node would be inaccessible via `MastNodeId`, as it is out of bounds, but because we truncated the length to `u32` when computing the node identifier, any request for the new node would get whatever node was associated with the truncated identifier).

The usage you linked in the AIR constraints code, I think is an example of a place where other code enforces that the length is < `u32::MAX` (at least, AFAIK) - if that's not actually the case, then the same issue applies there.

I would suggest that in any situation where the origin of the `usize` value is not crystal clear, we should use `try_into`, so as to catch bugs closer to the source. In cases where the `usize` has an obvious bounded range, then the cast via `as` is fine IMO.